src/monogame/vulkan/MGG_Vulkan.cpp

```diff
@@ -0,0 +1,973 @@
+};
+
+struct MGG_Buffer
+{
+	mgint length;
+};
+
+struct MGG_Texture
+{
+	MGTextureType type;
+	MGSurfaceFormat format;
+	mgint width;
+	mgint height;
+	mgint depth;
+	mgint mipmaps;
+	mgint slices;
+
+	mgbool isRenderTarget = false;
+	MGDepthFormat depthFormat = MGDepthFormat::None;
+	mgint multiSampleCount = 0;
+	MGRenderTargetUsage usage = MGRenderTargetUsage::PlatformContents;
+};
+
+struct MGG_InputLayout
+{
+	MGG_Shader* vertexShader;
+};
+
+struct MGG_Shader
+{
+	MGShaderStage stage;
+	mgbyte* bytecode;
+	mgint bytecodeSize;
+};
+
+struct MGG_BlendState
+{
+	MGG_BlendState_Info info[4];
+};
+
+struct MGG_DepthStencilState
+{
+	MGG_DepthStencilState_Info info;
+};
+
+struct MGG_RasterizerState
+{
+	MGG_RasterizerState_Info info;
+};
+
+struct MGG_SamplerState
+{
+	MGG_SamplerState_Info info;
+};
+
+
+struct MGG_System
+{
+	VkInstance instance;
+
+	std::vector<MGG_GraphicsAdapter*> adapters;
+};
+
+static MGG_System* mgg_system = nullptr;
+
+
+void MGG_EffectResource_GetBytecode(const char* name, mgbyte** bytecode, mgint* size)
+{
+	*bytecode = nullptr;
+	*size = 0;
+
+	// Get the handle of this DLL.
+	HMODULE module;
+	GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, (LPCSTR)&MGG_EffectResource_GetBytecode, &module);
+
+	LPCSTR id = nullptr;
+
+	if (strcmp(name, "AlphaTestEffect") == 0)
+		id = MAKEINTRESOURCEA(C_AlphaTestEffect);
+	else if (strcmp(name, "BasicEffect") == 0)
+		id = MAKEINTRESOURCEA(C_BasicEffect);
+	else if (strcmp(name, "DualTextureEffect") == 0)
+		id = MAKEINTRESOURCEA(C_DualTextureEffect);
+	else if (strcmp(name, "EnvironmentMapEffect") == 0)
+		id = MAKEINTRESOURCEA(C_EnvironmentMapEffect);
+	else if (strcmp(name, "SkinnedEffect") == 0)
+		id = MAKEINTRESOURCEA(C_SkinnedEffect);
+	else if (strcmp(name, "SpriteEffect") == 0)
+		id = MAKEINTRESOURCEA(C_SpriteEffect);
+
+	auto handle = ::FindResourceA(module, id, "BIN");
+	if (handle == nullptr)
+		return;
+
+	*size = SizeofResource(module, handle);
+	if (*size == 0)
+		return;
+
+	HGLOBAL global = LoadResource(module, handle);
+	if (global == nullptr)
+		return;
+
+	*bytecode = (mgbyte*)LockResource(global);
+}
+
+void System_GetAdapters(MGG_System* system)
+{
+	uint32_t count = 0;
+	VkResult res = vkEnumeratePhysicalDevices(system->instance, &count, NULL);
+	if (res != VK_SUCCESS)
+		return;
+
+	VkPhysicalDevice* gpus = (VkPhysicalDevice*)calloc(count, sizeof(*gpus));
+
+	res = vkEnumeratePhysicalDevices(system->instance, &count, gpus);
+	if (res == VK_SUCCESS)
+	{
+		for (uint32_t i = 0; i < count; ++i)
+		{
+			auto adapter = new MGG_GraphicsAdapter();
+			adapter->device = gpus[i];
+
+			vkGetPhysicalDeviceProperties(adapter->device, &adapter->properties);
+			vkGetPhysicalDeviceFeatures(adapter->device, &adapter->features);
+			vkGetPhysicalDeviceMemoryProperties(adapter->device, &adapter->memory);
+
+			system->adapters.push_back(adapter);
+		}
+	}
+
+	free((void*)gpus);
+}
+
+static MGG_System* System_Get()
+{
+	if (mgg_system != nullptr)
```

Comment:
I'm debating if we need some object like `MGG_System` exposed at the C# level to manage graphics state before a graphics device is created...  like initializing the Vulkan API (or initializing DX12).

Or should we let the native code manage this internally?

Reply:
Let's figure if we need this post-merge based on user requests.